### ISSUE TITLE
Add support for Oracle and SQL Server SYNONYMs in source code generation

### ISSUE THREAD

**Issue by lukaseder** [jOOQ/jOOQ#1563]
This is a sample query showing how to resolve transitive SYNONYM chains, in case this would be needed:

``` sql
SELECT "s3"."OWNER",
  "s3"."SYNONYM_NAME",
  connect_by_root "s3"."TABLE_OWNER" "TABLE_OWNER",
  connect_by_root "s3"."TABLE_NAME" "TABLE_NAME",
  SUBSTR(((((sys_connect_by_path((("s3"."TABLE_OWNER"
  || '.')
  || "s3"."TABLE_NAME"), ' <- ')
  || ' <- ')
  || "s3"."OWNER")
  || '.')
  || "s3"."SYNONYM_NAME"), 5)
FROM
  (SELECT "s2"."OWNER",
    "s2"."SYNONYM_NAME",
    "s2"."TABLE_OWNER",
    "s2"."TABLE_NAME"
  FROM (
    (SELECT "SYS"."ALL_SYNONYMS"."OWNER",
      "SYS"."ALL_SYNONYMS"."SYNONYM_NAME",
      "SYS"."ALL_SYNONYMS"."TABLE_OWNER",
      "SYS"."ALL_SYNONYMS"."TABLE_NAME"
    FROM "SYS"."ALL_SYNONYMS"
    )
  UNION
    (SELECT "SYS"."ALL_SYNONYMS"."OWNER",
      "SYS"."ALL_SYNONYMS"."SYNONYM_NAME",
      'PUBLIC',
      "SYS"."ALL_SYNONYMS"."TABLE_NAME"
    FROM "SYS"."ALL_SYNONYMS"
    )) "s2"
  WHERE ("s2"."OWNER", "s2"."SYNONYM_NAME") <> (("s2"."TABLE_OWNER", "s2"."TABLE_NAME"))
  ) "s3"
  START WITH EXISTS
  (SELECT 1 "one"
  FROM "SYS"."ALL_OBJECTS"
  WHERE ("s3"."TABLE_OWNER"              = "SYS"."ALL_OBJECTS"."OWNER"
  AND "s3"."TABLE_NAME"                  = "SYS"."ALL_OBJECTS"."OBJECT_NAME"
  AND "SYS"."ALL_OBJECTS"."OBJECT_TYPE" <> 'SYNONYM')
  )
  CONNECT BY ("s3"."TABLE_OWNER" = prior "s3"."OWNER"
AND "s3"."TABLE_NAME"            = prior "s3"."SYNONYM_NAME");
```

See this discussion here:
https://groups.google.com/d/topic/jooq-user/HXgEDfqJ2So/discussion

Update this answer here upon implementation:
http://stackoverflow.com/questions/11473358/can-jooq-generate-code-for-oracle-synonyms

The relevant documentation can be seen here:
http://docs.oracle.com/cd/B28359_01/server.111/b28286/statements_7001.htm

**Comment by rkettelerij** [comment]
+1

**Comment by lukaseder** [comment]
@rkettelerij : Thanks for the feedback. We've already been thinking about this a bit, and we're not yet 100% sure how to implement synonym support in generated code. We already resolve synonyms to their actual type, though. What is the use-case that you've had in mind?

**Comment by rkettelerij** [comment]
Synonyms are often used for doing cross-schema joins and query's. 

So you have a schema FOO and you want to represent objects (tables, views, procedures) from schema BAR in schema FOO. You can use an Oracle synonym like "create synonym foo.mytable for bar.othertable;" to accomplish this. As a result you only have to query schema FOO from your application while you're also getting data from schema BAR.

**Comment by lukaseder** [comment]
OK, I see. What we do right now (what I meant by resolving synonyms) is we may have an object type `FOO.BOOK_T`, and a procedure `BAR.PKG.GET_BOOKS` that returns a synonym `BAR.BOOK_T`. Right now, we resolve `BAR.BOOK_T` to be `FOO.BOOK_T` and generate a procedure that pretends that there is no synonym.

This makes sense with types and procedures.

What you have in mind (and what this issue is about) is that if there is a synonym `foo.mytable` for `bar.othertable`, then all the generated artefacts will be duplicated, such that you can write `MYTABLE` in Java, when you really mean `OTHERTABLE`, as well.

The difficult thing here is to get generated foreign key relationships right but I guess we can find a way...

**Comment by rkettelerij** [comment]
> What you have in mind (and what this issue is about) is that if there is a synonym foo.mytable for bar.othertable, then all the generated artefacts will be duplicated, such that you can write MYTABLE in Java, when you really mean OTHERTABLE, as well.

Right, that's what I mean.

**Comment by jmhernandez127** [comment]
any news on the support for synonyms @lukaseder ? thanks in advance

**Comment by lukaseder** [comment]
There are no news as you can see 😉 Why don't you provide news? What's your expectation here? What would you like jOOQ to do?

**Comment by jmhernandez127** [comment]
> There are no news as you can see 😉 Why don't you provide news? What's your expectation here? What would you like jOOQ to do?

No worries @lukaseder. Just similar to what @rkettelerij was trying to achieve. But we have a package that has a synonym we need to use since we have a 2 different packages used by different clients but pointing to one synonym name only for core functionality purposes. 

But I guess we have an idea how to do a workaround on this for now, hopefully it will not be that complicated and will solve our issue. We were thinking of retrieving the actual object (package) name of the synonym and use the query to generate a custom java generated package out of the actual package names and sort of mimicking the synonym or "aliasing" the actual package name being recognized by jooq.

**Comment by lukaseder** [comment]
> We were thinking of retrieving the actual object (package) name of the synonym and use the query to generate a custom java generated package out of the actual package names and sort of mimicking the synonym or "aliasing" the actual package name being recognized by jooq.

Yes, that's what I had just thought myself. Sounds like a job for codegen-time or runtime schema mapping:

- https://www.jooq.org/doc/latest/manual/code-generation/codegen-advanced/codegen-config-database/codegen-database-catalog-and-schema-mapping/
- https://www.jooq.org/doc/latest/manual/sql-building/dsl-context/custom-settings/settings-render-mapping/

I see how synonyms can be used to "import" objects between schemas. But I do wonder if this is really worth the trouble, supporting them in jOOQ this way can complicate a lot of things. For example:

- A table synonym can have foreign keys or check constraints pointing to actual tables from another schema. Should we resolve those, transitively or not?
- A table synonym can have foreign keys or check constraints pointing to other table synonyms.

This is by no means a simple thing to implement correctly. We already resolve synonyms in the code generator in the latter case, because we need to know what *actual* object a constraint (or whatever) is pointing to as the dictionary views produce the synonym, not the actual name (and we then pretend the synonym doesn't exist). In most cases, the synonym name is not the desirable name, people want jOOQ's code generator to resolve synonyms to the actual object.

This use-case here is a use-case where the synonym should be resolved only to find out the type of the actual object, but then be left alone to produce a copy of it. But what of all the references to the synonym that is now no longer resolved? It gets much more complicated with UDTs. What if a package procedure references an object type by synonym name? What if an object type references an attribute's nested object type by synonym name? Would you want the synonym name to appear in your generated code? But the type name that is used to create UDTs using PL/SQL blocks would have to remain the original one, I'd suspect?

Perhaps there's a way to resolve all this, but the cost/benefit ratio of this implementation isn't very good, given the amount of possible bugs this could produce. Even among the Oracle users, only few use synonyms, *and* would like their synonyms *not* to be resolved in generated code.

**Comment by lukaseder** [comment]
Duplicate of https://github.com/jOOQ/jOOQ/issues/9574 (implemented in jOOQ 3.20)
